tests/cpp/transforms/make_block_local_test.cpp

```diff
@@ -130,7 +131,10 @@ TEST_F(MakeBlockLocalTest, Basic) {
   // the generated BLS offset is correct for the given global loop indices.
   for (int idx0 = 0; idx0 < loop_shape_at(0); ++idx0) {
     for (int idx1 = 0; idx1 < loop_shape_at(1); ++idx1) {
-      const int loop_indices_vals[] = {idx0, idx1};
+      const int loop_indices_vals[] = {
+          get_block_size(0) * idx0,
```

Comment:
IIUC, without the trailing_bits, the `loop_indices_vals` will not be scaled by block_size? (I.e., the looping range is the same as the shape of `struct_for_snode_`)? The block corner itself is not scaled by block_size, either. Instead, `make_bls()` will generate a statement to scale that by coefficients.

Reply:
Yep. exactly right :)

Reply:
Then neither L135-136 nor  `get_block_corner()` needs to multiply by `get_block_size(x)`? I guess we also need to update how `expected_bls_offset_in_bytes` are calculated..

Reply:
Oh sorry I misread your comment. So the coefficients of the scratch pad come directly from the user multiplying the loop index with the block size. This is why in the mpm example, we have to manually rescale the loop index to make it a block_corner before passing it to assume_in_range. 
Also, This test is not a perfect copy of the of mpm example since we are not using the position of the particles to get the central index of the neighborhood to do the scatter/gather. Instead we are just using the loop index, which is just a block index of the grid snode. So we need to rescale this one too, making it also simply a block corner.

Reply:
> So the coefficients of the scratch pad come directly from the user multiplying the loop index with the block size. This is why in the mpm example, we have to manually rescale the loop index to make it a block_corner before passing it to assume_in_range.

Right, I should have better explained the test code: the `loop_indices_vals` array is meant to mock the value of `LoopIndexStmt` for each axis in the `ArithmeticInterpretor`. Similarly, `block_corner_vals` is to mock each `BlockCornerIndexStmt` (see https://github.com/taichi-dev/taichi/blob/22e33a605b5caccc62a02be491a3dc33c2f31c91/tests/cpp/transforms/make_block_local_test.cpp#L170-L177). As you've pointed out, it is the user that is manually multiplying loop index by the block size (L105-108). So `LoopIndexStmt`/`BlockCornerIndexStmt` themselves should not be scaled yet.

We happen to use `loop_indices_vals` and `block_corner_vals` to compute `expected_bls_offset_in_bytes`, which is probably causing lots of confusion. Since the formula of BLS index <--> global indices  has changed, I think we can update how `expected_bls_offset_in_bytes` gets computed. I.e.:

```
expected_bls_offset_in_bytes = sum_i inc_i
inc_i = ( loop_index_i * block_size   # multiplication done by the user
            - block_corner_i * block_size  # multiplication generated by the compiler
            - bls_lower_bounds_i)
```

WDYT?

Reply:
Ahh yes, I think I see what you mean. the BlockCornerIndexStmt _should_ be returning consecutive indices in this cases since we are looping over a dynamic snode which has zero num_bits for i and j. So it does not get refined to non conecutive indices.

Reply:
I have implemented your suggestion